File: hw1-kyle-pish/p4/p4_unit_test.cpp
// Header comment
// Test out the functions in p4_lib.py
// Using the Catch C++ unit testing framework: https://github.com/catchorg/Catch2/tree/v2.x

#define CATCH_CONFIG_MAIN  // This tells Catch to provide a main() - only do this in one cpp file
#include "catch.hpp"

#include "p4_lib.h"


TEST_CASE( "Testing all p4_lib functions") {
    auto jan1 = std::make_pair(1, 1);
    auto jan2 = std::make_pair(1, 2);
    auto feb28 = std::make_pair(2, 28);
    auto dec1 = std::make_pair(12, 1);
    auto dec31 = std::make_pair(12, 31);

    SECTION( "date_is_before" ) {
        REQUIRE(date_is_before(jan1, jan2));
        REQUIRE(date_is_before(jan2, feb28));
        REQUIRE(date_is_before(jan2, dec1));
        REQUIRE(date_is_before(feb28, dec31));
    }

    SECTION( "days_in_month" ) {
        REQUIRE(days_in_month(1) == 31);
        REQUIRE(days_in_month(2) == 28);
        REQUIRE(days_in_month(12) == 31);
    }

    SECTION( "date_plus" ) {
        REQUIRE(date_plus(jan1, 1) == jan2);
        REQUIRE(date_plus(feb28, 1) == std::make_pair(3, 1));
        REQUIRE(date_plus(dec31, 2) == jan2);
    }

    SECTION( "latest_date" ) {
        REQUIRE(latest_date({jan1, jan2}) == jan2);
        REQUIRE(latest_date({feb28, jan1, jan2}) == feb28);
        REQUIRE(latest_date({jan1, feb28, dec31, jan2}) == dec31);
        REQUIRE(latest_date({jan1, feb28, dec1, jan2}) == dec1);
    }
}

File: hw1-kyle-pish/p4/p4_unit_test.py
# Header
# Test out the functions in p4_lib.py
# Using the unittest library in Python's standard library

from p4_lib import date_is_before, days_in_month, date_plus, latest_date

import unittest


class LibTestCase(unittest.TestCase):
    def setUp(self):
        self.jan1 = (1, 1)
        self.jan2 = (1, 2)
        self.feb28 = (2, 28)
        self.dec1 = (12, 1)
        self.dec31 = (12, 31)

    def test_date_is_before(self):
        self.assertTrue(date_is_before(self.jan1, self.jan2))
        self.assertTrue(date_is_before(self.jan2, self.feb28))
        self.assertTrue(date_is_before(self.jan2, self.dec1))
        self.assertTrue(date_is_before(self.feb28, self.dec31))

    def test_days_in_month(self):
        self.assertEqual(days_in_month(1), 31)
        self.assertEqual(days_in_month(2), 28)
        self.assertEqual(days_in_month(12), 31)

    def test_date_plus(self):
        self.assertEqual(date_plus(self.jan1, 1), self.jan2)
        self.assertEqual(date_plus(self.feb28, 1), (3, 1))
        self.assertEqual(date_plus(self.dec31, 2), self.jan2)

    def test_latest_date(self):
        self.assertEqual(latest_date([self.jan1, self.jan2]), self.jan2)
        self.assertEqual(latest_date([self.feb28, self.jan1, self.jan2]), self.feb28)
        self.assertEqual(latest_date([self.jan1, self.feb28, self.dec31, self.jan2]), self.dec31)
        self.assertEqual(latest_date([self.jan1, self.feb28, self.dec1, self.jan2]), self.dec1)


if __name__ == '__main__':
    unittest.main()
